test/plugins/optional-auth/src/main/java/org/elasticsearch/test/plugin/optionalauth/OptionalAuthenticationDemandingActionFilter.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.test.plugin.optionalauth;
+
+import org.elasticsearch.ElasticsearchSecurityException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.support.ActionFilter;
+import org.elasticsearch.action.support.ActionFilterChain;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.rest.RestController;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+
+/**
+ * Filter that demands the "Authorization" header if the "demain-auth" header is sent. The only valid authorization is http basic with
+ * "Aladdin" as the username and "open sesame" as the password.
+ */
+public class OptionalAuthenticationDemandingActionFilter implements ActionFilter {
+    /**
+     * The authorization required. Corresponds to username="Aladdin" and password="open sesame". It is the example in
+     * <a href="https://tools.ietf.org/html/rfc1945#section-11.1">HTTP/1.0's RFC</a>.
+     */
+    private static final String REQUIRED_AUTH = "Basic QWxhZGRpbjpvcGVuIHNlc2FtZQ==";
+    private static final String DEMAND_HEADER = "demand-auth";
+    private static final String AUTHORIZATION_HEADER = "Authorization";
+    private final ThreadContext context;
+
+    @Inject
+    public OptionalAuthenticationDemandingActionFilter(ThreadPool threadPool, RestController controller) {
+        context = threadPool.getThreadContext();
+        controller.registerRelevantHeaders(DEMAND_HEADER, AUTHORIZATION_HEADER);
+    }
+
+    @Override
+    public int order() {
+        return Integer.MIN_VALUE;
+    }
+
+    private void checkAuth() {
+        if (context.getHeader(DEMAND_HEADER) == null) {
```

Comment:
why do we need a plugin with optional auth? Can we just require basic auth whenever the plugin is installed? I guess I don't love this header trick, and I wonder why it is necessary.